### ISSUE TITLE
[Feature]: Faster guided decoding for pre-defined output

### ISSUE THREAD

**Issue by captify-sivakhno** [vllm-project/vllm#8075]
### 🚀 The feature, motivation and pitch

vLLM provides a guided generation via outlines, e.g. 
```Python
from outlines import models, generate
model = models.VLLM(llm)
generator = generate.choice(model, array_of_categories)
```

This however does not fully utilise deterministic nature of output. Since categories are pre-defined, I wander if decoding can run in parallel for all tokens to increase throughput of decoding. 
I would be interested in trying to make contributions if folks can provide guidance on steps and where to look in vLLM codebase.

### Alternatives

Speculative generation is somewhat similar, for example when ngrams are used in place of approximation model. 

### Before submitting a new issue...

- [X] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.
